cvexporter/exporter.py

```diff
@@ -0,0 +1,129 @@
+                g.add_metric([ns,
+                             context,
+                             service,
+                             container,
+                             sev],
+                             vulncount)
+        yield g
+
+
+def merge_metrics(data):
+    checked = set()
+    merged_data = []
+    for ns, sc, ss, cn, sev, val in set(data):
+        if not (ns, sc, ss, cn, sev) in checked:
+            checked.add((ns, sc, ss, cn, sev))
+            vtotal = 0
+            for ns2, sc2, ss2, cn2, sev2, val2 in data:
+                if (ns2 == ns and sc2 == sc and ss2 == ss and
+                        cn2 == cn and sev2 == sev):
+                    vtotal = vtotal + val2
+            merged_data.append((ns, sc, ss, cn, sev, vtotal))
+    return merged_data
+
+
+@lc()
+def collect_metrics():
+    metric_data = []
+    for ns, context, service, container, imageurl in \
+            openshift_cluster.get_images(TARGET_NS, SKIP_NS):
+        if quay.from_quay(imageurl):
+            vulnerabilities = quay.get_vulnerabilities(imageurl)
+            if vulnerabilities is not None:
+                for sev in quay.severities():
+                    metric_data.append((ns,
+                                        context,
+                                        service,
+                                        container,
+                                        sev,
+                                        vulnerabilities[sev]))
+        else:
+            #  logging.info(f'(n:{ns}) {pod}/{container} is not sourced ' +
+            #               ' from Quay.io')
+            continue
+    return merge_metrics(metric_data)
+
+
+@lc()
+def metrics_cache():
+    if collect_metrics.cache_info().currsize > 0:
+        return collect_metrics()
+
+
+def cache_utilization(cache_stats):
+    return ('Cache (size:{}, calls:{}, cache_usage:{:.0%})'
+            .format(cache_stats.currsize,
+                    cache_stats.hits + cache_stats.misses,
+                    (cache_stats.hits /
+                     (cache_stats.hits+cache_stats.misses))))
+
+
+def balance_caches():
+    logging.info('Quay Vulnerabilities {}'
+                 .format(cache_utilization(
+                  quay.wget_vulnerabilities.cache_info())))
+    quay.wget_vulnerabilities.cache_clear()
+    logging.info('Openshift Resources {}'
+                 .format(cache_utilization(
+                  openshift_cluster.get_resources_by_kind.cache_info())))
+    openshift_cluster.get_resources_by_kind.cache_clear()
+
+
+if __name__ == '__main__':
+    logging.basicConfig(format='%(levelname)s: %(message)s',
+                        level=getattr(logging, LOG_LEVEL))
+    logging.info('Initializing exporter')
+    prom.start_http_server(LISTEN_PORT)
+    REGISTRY.register(CveMetrics())
+
+    logging.info('Entering collection loop')
+    logging.info('Note: Missing label notices will only appear once')
+    while True:
+        collect_metrics.cache_clear()
```

Comment:
I think we sholdn't use caches here. https://gist.github.com/jmelis/259643c545c5e3e407a162be39f311ef

Reply:
changed